Libraries/Connect/Internal/Unary/UnaryAsyncWrapper.swift

```
@@ -41,20 +41,33 @@ actor UnaryAsyncWrapper<Output: ProtobufMessage>: Sendable {
     ///
     /// - returns: The response/result of the request.
     func send() async -> ResponseMessage<Output> {
-        return await withTaskCancellationHandler(operation: {
-            return await withCheckedContinuation { continuation in
-                if Task.isCancelled {
+        await withTaskCancellationHandler {
+            await withCheckedContinuation { continuation in
+                guard !Task.isCancelled else {
                     continuation.resume(
-                        returning: .init(code: .canceled, result: .failure(.canceled()))
+                        returning: ResponseMessage(
+                            code: .canceled,
+                            result: .failure(.canceled())
+                        )
                     )
-                } else {
-                    self.cancelable = self.sendUnary { response in
-                        continuation.resume(returning: response)
+                    return
+                }
+
+                let hasResumed = Locked(false)
+                self.cancelable = self.sendUnary { response in
+                    guard !hasResumed.value else {
+                        return assertionFailure("Attempting to resume continuation twice")
```

What's your thought process for keeping this assertion in place? I assume it will trigger in the cases described by the PR description, which could be disruptive for local development

---

My reasoning was mostly that I felt like failing silently didn't really seem appropriate either.

In the interest of time, the change here specifically addresses the continuation resuming twice, which was resulting in the worst possible experience in both debug and release - a crash. It does not necessarily address the root cause for _why_ `sendUnary`'s closure is getting triggered a second time.

It felt like keeping an assertion there, such that the debugger pauses when it happens during development, would provide a means to re-raise the issue and prompt deeper analysis of the root cause at a later date if we see it happening regularly.

Up to you how you'd like to proceed though. I can just as easily `os_log` something, create an issue and put it in a comment, etc.

---

> It felt like keeping an assertion there, such that the debugger pauses when it happens during development, would provide a means to re-raise the issue and prompt deeper analysis of the root cause at a later date if we see it happening regularly.
>
> Up to you how you'd like to proceed though. I can just as easily os_log something, create an issue and put it in a comment, etc.

I get where you're coming from. That said, if I'm understanding correctly this is an issue in the underlying library/dependency - not in how the caller is using the API. If a user hits an assertion in debug I think they'd think _they_ were doing something wrong, but in practice it doesn't seem like they need to do anything. Since they end up getting a non-successful response once anyway, I think adding an assertion will introduce confusion/distraction to the end user. IMO keeping an `os_log` here seems sufficient unless you feel strongly

---

Works for me! Would you also like us to make an Issue just to track the underlying problem for later investigation?

---

I think an inline comment is probably sufficient :)